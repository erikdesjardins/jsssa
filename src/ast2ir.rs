use std::collections::HashMap;

use ast;
use ir;
use util::*;

type ScopeMap = Map<String, ir::Ref<ir::Mutable>>;

pub fn convert(ast: ast::File) -> ir::Block {
    let ast::File {
        program:
            ast::Program {
                body,
                directives,
                source_type: _,
            },
    } = ast;

    convert_block(body, directives, &HashMap::default())
}

fn convert_block(
    body: Vec<ast::Statement>,
    directives: Vec<ast::Directive>,
    parent_scopes: &ScopeMap,
) -> ir::Block {
    let mut scope = parent_scopes.clone();

    let children = body.into_iter()
        .flat_map(|stmt| convert_statement(stmt, &mut scope))
        .collect();

    ir::Block {
        directives: directives.into_iter().map(|d| d.value.value).collect(),
        children,
    }
}

fn convert_statement(stmt: ast::Statement, scope: &mut ScopeMap) -> Vec<ir::Stmt> {
    use ast::Statement::*;

    match stmt {
        ExpressionStatement(ast::ExpressionStatement { expression }) => {
            let (mut exprs, last_expr) = convert_expression(expression, scope);
            exprs.push(ir::Stmt::Expr(last_expr));
            exprs
        },
        BlockStatement(ast::BlockStatement { body, directives }) =>
            vec![ir::Stmt::Block(box convert_block(body, directives, scope))],
        EmptyStatement(ast::EmptyStatement {}) =>
            vec![],
        DebuggerStatement(ast::DebuggerStatement {}) =>
            vec![ir::Stmt::Debugger],
        WithStatement(_) =>
            unimplemented!("with() statement not supported"),
        ReturnStatement(ast::ReturnStatement { argument }) => {
            let ref_ = ir::Ref::new("return_".to_string());
            match argument {
                Some(argument) => {
                    let (mut exprs, return_value) = convert_expression(argument, scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    exprs
                },
                None => {
                    vec![
                        ir::Stmt::Assign(ref_.clone(), ir::Expr::Undefined),
                        ir::Stmt::Return(ref_),
                    ]
                }
            }
        },
        LabeledStatement(_) |
        BreakStatement(_) |
        ContinueStatement(_) =>
            unimplemented!("labels not yet supported"),
        IfStatement(ast::IfStatement { test, consequent, alternate }) => {
            let ref_ = ir::Ref::new("if_".to_string());
            let (mut exprs, test_value) = convert_expression(test, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), test_value));
            exprs.push(ir::Stmt::IfElse(
                ref_,
                {
                    let children = convert_statement(*consequent, scope);
                    box ir::Block { directives: vec![], children }
                },
                match alternate {
                    Some(alternate) => {
                        let children = convert_statement(*alternate, scope);
                        box ir::Block { directives: vec![], children }
                    },
                    None => box ir::Block::empty()
                },
            ));
            exprs
        },
        SwitchStatement(_) =>
            // remember, switch cases aren't evaluated (!) until they're checked for equality
            unimplemented!("switch statements not yet supported"),
        ThrowStatement(ast::ThrowStatement { argument }) => {
            let ref_ = ir::Ref::new("throw_".to_string());
            let (mut exprs, throw_value) = convert_expression(argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), throw_value));
            exprs.push(ir::Stmt::Throw(ref_));
            exprs
        },
        TryStatement(ast::TryStatement { block, handler, finalizer }) => {
            let try = ir::Stmt::Try(
                {
                    let ast::BlockStatement { body, directives } = block;
                    box convert_block(body, directives, scope)
                },
                handler.map(|ast::CatchClause { param, body }| {
                    let ast::BlockStatement { body, directives } = body;
                    let mut catch_scope = scope.clone();
                    let ref_ = convert_pattern(param, &mut catch_scope);
                    (ref_, box convert_block(body, directives, &catch_scope))
                }),
                finalizer.map(|finalizer| {
                    let ast::BlockStatement { body, directives } = finalizer;
                    box convert_block(body, directives, scope)
                }),
            );
            vec![try]
        },
        _ => unimplemented!(),
    }
}

fn convert_expression(expr: ast::Expression, scope: &ScopeMap) -> (Vec<ir::Stmt>, ir::Expr) {
    use ast::Expression::*;

    match expr {
        Identifier(ast::Identifier { name }) => {
            let expr = match scope.get(&name) {
                Some(ref_) => ir::Expr::ReadBinding(ref_.clone()),
                None => ir::Expr::ReadGlobal(name),
            };
            (vec![], expr)
        }
        RegExpLiteral(ast::RegExpLiteral { pattern, flags }) => {
            (vec![], ir::Expr::RegExp(pattern, flags))
        }
        NullLiteral(ast::NullLiteral {}) => (vec![], ir::Expr::Null),
        StringLiteral(ast::StringLiteral { value }) => (vec![], ir::Expr::String(value)),
        BooleanLiteral(ast::BooleanLiteral { value }) => (vec![], ir::Expr::Bool(value)),
        NumericLiteral(ast::NumericLiteral { value }) => (vec![], ir::Expr::Number(value)),
        ThisExpression(ast::ThisExpression {}) => (vec![], ir::Expr::This),
        ArrowFunctionExpression(ast::ArrowFunctionExpression {
            params,
            body,
            async,
        }) => {
            use ast::BlockStmtOrExpr::*;

            let mut fn_scope = scope.clone();
            let refs = params
                .into_iter()
                .map(|param| convert_pattern(param, &mut fn_scope))
                .collect();
            let body = match *body {
                BlockStatement(block) => {
                    let ast::BlockStatement { body, directives } = block;
                    convert_block(body, directives, &fn_scope)
                }
                Expression(expr) => {
                    let ref_ = ir::Ref::new("arrow_".to_string());
                    let (mut exprs, return_value) = convert_expression(expr, &fn_scope);
                    exprs.push(ir::Stmt::Assign(ref_.clone(), return_value));
                    exprs.push(ir::Stmt::Return(ref_));
                    ir::Block {
                        directives: vec![],
                        children: exprs,
                    }
                }
            };
            let func = ir::Expr::Function(ir::FnKind::Arrow { async }, None, refs, box body);
            (vec![], func)
        }
        YieldExpression(ast::YieldExpression { argument, delegate }) => {
            let ref_ = ir::Ref::new("yield_".to_string());
            let (mut exprs, yield_value) = match argument {
                Some(argument) => convert_expression(*argument, scope),
                None => (vec![], ir::Expr::Undefined),
            };
            exprs.push(ir::Stmt::Assign(ref_.clone(), yield_value));
            let kind = if delegate {
                ir::YieldKind::Delegate
            } else {
                ir::YieldKind::Single
            };
            (exprs, ir::Expr::Yield(kind, ref_))
        }
        AwaitExpression(ast::AwaitExpression { argument }) => {
            let ref_ = ir::Ref::new("await_".to_string());
            let (mut exprs, await_value) = convert_expression(*argument, scope);
            exprs.push(ir::Stmt::Assign(ref_.clone(), await_value));
            (exprs, ir::Expr::Await(ref_))
        }
        _ => unimplemented!(),
    }
}

fn convert_pattern(pat: ast::Pattern, scope: &mut ScopeMap) -> ir::Ref<ir::Mutable> {
    use ast::Pattern::*;

    match pat {
        Identifier(ast::Identifier { name }) => {
            let ref_ = ir::Ref::new(name.clone());
            scope.insert(name, ref_.clone());
            ref_
        }
        MemberExpression(_) | ObjectPattern(_) | ArrayPattern(_) | RestElement(_)
        | AssignmentPattern(_) => unimplemented!("complex patterns not yet supported"),
    }
}
