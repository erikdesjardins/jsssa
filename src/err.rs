use std::fmt::{self, Debug, Display};

pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;

/// Delegates `Debug` to `Display`, since `Termination` always uses `Debug` to print errors
pub struct NiceError(Error);

impl Debug for NiceError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        Display::fmt(&self.0, f)
    }
}

impl<T: Into<Error>> From<T> for NiceError {
    fn from(display: T) -> Self {
        NiceError(display.into())
    }
}
