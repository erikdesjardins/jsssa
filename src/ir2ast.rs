use swc_common::Span;
use swc_ecma_ast as ast;

use crate::ir;
use crate::ir::scope;
use crate::utils::P;

pub fn convert(ir: ir::Block) -> ast::Script {
    // todo perform inlining at this stage? (i.e. scan backwards for all usages)
    let body = convert_block(ir, &scope::Ir::default());
    ast::Script {
        span: span(),
        body,
        shebang: None,
    }
}

fn convert_block(block: ir::Block, parent_scope: &scope::Ir) -> Vec<ast::Stmt> {
    let mut scope = parent_scope.clone();

    let ir::Block { children } = block;

    children
        .into_iter()
        .map(|stmt| convert_stmt(stmt, &mut scope))
        .collect()
}

fn convert_stmt(stmt: ir::Stmt, scope: &mut scope::Ir) -> ast::Stmt {
    match stmt {
        ir::Stmt::Expr { target, expr } => {
            let expr = convert_expr(expr, scope);
            if target.maybe_used() {
                let name = scope.declare_ssa(target);
                ast::Stmt::Decl(ast::Decl::Var(ast::VarDecl {
                    span: span(),
                    kind: ast::VarDeclKind::Var,
                    decls: vec![ast::VarDeclarator {
                        span: span(),
                        name: ast::Pat::Ident(ast::Ident {
                            span: span(),
                            sym: name,
                            type_ann: None,
                            optional: false,
                        }),
                        init: Some(P(expr)),
                        definite: false,
                    }],
                    declare: false,
                }))
            } else {
                ast::Stmt::Expr(P(expr))
            }
        }
        ir::Stmt::DeclareMutable { target, val } => {
            let name = scope.declare_mutable(target);
            ast::Stmt::Decl(ast::Decl::Var(ast::VarDecl {
                span: span(),
                kind: ast::VarDeclKind::Var,
                decls: vec![ast::VarDeclarator {
                    span: span(),
                    name: ast::Pat::Ident(ast::Ident {
                        span: span(),
                        sym: name,
                        type_ann: None,
                        optional: false,
                    }),
                    init: Some(P(read_ssa_to_expr(val, scope))),
                    definite: false,
                }],
                declare: false,
            }))
        }
        ir::Stmt::WriteMutable { target, val } => match scope.get_mutable(&target) {
            Some(existing_name) => ast::Stmt::Expr(P(ast::Expr::Assign(ast::AssignExpr {
                span: span(),
                op: ast::AssignOp::Assign,
                left: ast::PatOrExpr::Pat(P(ast::Pat::Ident(ast::Ident {
                    span: span(),
                    sym: existing_name,
                    type_ann: None,
                    optional: false,
                }))),
                right: P(read_ssa_to_expr(val, scope)),
            }))),
            None => unreachable!("writing to undeclared mutable ref"),
        },
        ir::Stmt::WriteGlobal { target, val } => {
            ast::Stmt::Expr(P(ast::Expr::Assign(ast::AssignExpr {
                span: span(),
                op: ast::AssignOp::Assign,
                left: ast::PatOrExpr::Pat(P(ast::Pat::Ident(ast::Ident {
                    span: span(),
                    sym: target,
                    type_ann: None,
                    optional: false,
                }))),
                right: P(read_ssa_to_expr(val, scope)),
            })))
        }
        ir::Stmt::WriteMember { obj, prop, val } => {
            ast::Stmt::Expr(P(ast::Expr::Assign(ast::AssignExpr {
                span: span(),
                op: ast::AssignOp::Assign,
                left: ast::PatOrExpr::Pat(P(ast::Pat::Expr(P(ast::Expr::Member(
                    ast::MemberExpr {
                        span: span(),
                        obj: ast::ExprOrSuper::Expr(P(read_ssa_to_expr(obj, scope))),
                        prop: P(read_ssa_to_expr(prop, scope)),
                        computed: true,
                    },
                ))))),
                right: P(read_ssa_to_expr(val, scope)),
            })))
        }
        ir::Stmt::Return { val } => ast::Stmt::Return(ast::ReturnStmt {
            span: span(),
            arg: Some(P(read_ssa_to_expr(val, scope))),
        }),
        ir::Stmt::Throw { val } => ast::Stmt::Throw(ast::ThrowStmt {
            span: span(),
            arg: P(read_ssa_to_expr(val, scope)),
        }),
        ir::Stmt::Break => ast::Stmt::Break(ast::BreakStmt {
            span: span(),
            label: None,
        }),
        ir::Stmt::Continue => ast::Stmt::Continue(ast::ContinueStmt {
            span: span(),
            label: None,
        }),
        ir::Stmt::Debugger => ast::Stmt::Debugger(ast::DebuggerStmt { span: span() }),
        ir::Stmt::Block { body } => ast::Stmt::Block(ast::BlockStmt {
            span: span(),
            stmts: convert_block(*body, scope),
        }),
        ir::Stmt::Loop { body } => ast::Stmt::While(ast::WhileStmt {
            span: span(),
            test: P(ast::Expr::Lit(ast::Lit::Bool(ast::Bool {
                span: span(),
                value: true,
            }))),
            body: P(ast::Stmt::Block(ast::BlockStmt {
                span: span(),
                stmts: convert_block(*body, scope),
            })),
        }),
        ir::Stmt::ForEach {
            kind,
            init,
            mut body,
        } => {
            let mut for_scope = scope.clone();
            let var_name = match body.children.get(0) {
                Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Argument { index: 0 },
                }) => {
                    let name = for_scope.declare_ssa(target.clone());
                    // drop this expr
                    body.children.remove(0);
                    name
                }
                _ => for_scope.declare_ssa(ir::Ref::dead()),
            };
            let left = ast::VarDeclOrPat::VarDecl(ast::VarDecl {
                span: span(),
                kind: ast::VarDeclKind::Var,
                decls: vec![ast::VarDeclarator {
                    span: span(),
                    name: ast::Pat::Ident(ast::Ident {
                        span: span(),
                        sym: var_name,
                        type_ann: None,
                        optional: false,
                    }),
                    init: None,
                    definite: false,
                }],
                declare: false,
            });
            let right = P(read_ssa_to_expr(init, scope));
            let body = P(ast::Stmt::Block(ast::BlockStmt {
                span: span(),
                stmts: convert_block(*body, &for_scope),
            }));
            match kind {
                ir::ForKind::In => ast::Stmt::ForIn(ast::ForInStmt {
                    span: span(),
                    left,
                    right,
                    body,
                }),
                ir::ForKind::Of => ast::Stmt::ForOf(ast::ForOfStmt {
                    span: span(),
                    await_token: None,
                    left,
                    right,
                    body,
                }),
            }
        }
        ir::Stmt::IfElse { cond, cons, alt } => ast::Stmt::If(ast::IfStmt {
            span: span(),
            test: P(read_ssa_to_expr(cond, scope)),
            cons: P(ast::Stmt::Block(ast::BlockStmt {
                span: span(),
                stmts: convert_block(*cons, scope),
            })),
            alt: Some(P(ast::Stmt::Block(ast::BlockStmt {
                span: span(),
                stmts: convert_block(*alt, scope),
            }))),
        }),
        ir::Stmt::Try {
            body,
            catch,
            finally,
        } => ast::Stmt::Try(ast::TryStmt {
            span: span(),
            block: ast::BlockStmt {
                span: span(),
                stmts: convert_block(*body, scope),
            },
            handler: catch.map(|mut body| {
                let mut catch_scope = scope.clone();
                let param_name = match body.children.get(0) {
                    Some(ir::Stmt::Expr {
                        target,
                        expr: ir::Expr::Argument { index: 0 },
                    }) => {
                        let name = catch_scope.declare_ssa(target.clone());
                        // drop this expr
                        body.children.remove(0);
                        Some(name)
                    }
                    _ => None,
                };
                ast::CatchClause {
                    span: span(),
                    param: param_name.map(|param_name| {
                        ast::Pat::Ident(ast::Ident {
                            span: span(),
                            sym: param_name,
                            type_ann: None,
                            optional: false,
                        })
                    }),
                    body: ast::BlockStmt {
                        span: span(),
                        stmts: convert_block(*body, &catch_scope),
                    },
                }
            }),
            finalizer: finally.map(|body| ast::BlockStmt {
                span: span(),
                stmts: convert_block(*body, scope),
            }),
        }),
    }
}

fn convert_expr(expr: ir::Expr, scope: &scope::Ir) -> ast::Expr {
    match expr {
        ir::Expr::Bool { value } => ast::Expr::Lit(ast::Lit::Bool(ast::Bool {
            span: span(),
            value,
        })),
        ir::Expr::Number { value } => ast::Expr::Lit(ast::Lit::Num(ast::Number {
            span: span(),
            value,
        })),
        ir::Expr::String { value, has_escape } => ast::Expr::Lit(ast::Lit::Str(ast::Str {
            span: span(),
            value,
            has_escape,
        })),
        ir::Expr::Null => ast::Expr::Lit(ast::Lit::Null(ast::Null { span: span() })),
        ir::Expr::Undefined => ast::Expr::Unary(ast::UnaryExpr {
            span: span(),
            op: ast::UnaryOp::Void,
            arg: P(ast::Expr::Lit(ast::Lit::Num(ast::Number {
                span: span(),
                value: 0.0,
            }))),
        }),
        ir::Expr::This => ast::Expr::This(ast::ThisExpr { span: span() }),
        ir::Expr::Read { source } => read_ssa_to_expr(source, scope),
        ir::Expr::ReadMutable { source } => {
            let name = match scope.get_mutable(&source) {
                Some(name) => name,
                None => unreachable!("reading from undeclared mutable ref"),
            };
            ast::Expr::Ident(ast::Ident {
                span: span(),
                sym: name,
                type_ann: None,
                optional: false,
            })
        }
        ir::Expr::ReadGlobal { source } => ast::Expr::Ident(ast::Ident {
            span: span(),
            sym: source,
            type_ann: None,
            optional: false,
        }),
        ir::Expr::ReadMember { obj, prop } => ast::Expr::Member(ast::MemberExpr {
            span: span(),
            obj: ast::ExprOrSuper::Expr(P(read_ssa_to_expr(obj, scope))),
            prop: P(read_ssa_to_expr(prop, scope)),
            computed: true,
        }),
        ir::Expr::Array { elems } => ast::Expr::Array(ast::ArrayLit {
            span: span(),
            elems: elems
                .into_iter()
                .map(|elem| {
                    elem.map(|(kind, val)| ast::ExprOrSpread {
                        spread: match kind {
                            ir::EleKind::Single => None,
                            ir::EleKind::Spread => Some(span()),
                        },
                        expr: P(read_ssa_to_expr(val, scope)),
                    })
                })
                .collect(),
        }),
        ir::Expr::Object { props } => ast::Expr::Object(ast::ObjectLit {
            span: span(),
            props: props
                .into_iter()
                .map(|(kind, prop, val)| {
                    ast::PropOrSpread::Prop(P(match kind {
                        ir::PropKind::Simple => ast::Prop::KeyValue(ast::KeyValueProp {
                            key: ast::PropName::Computed(P(read_ssa_to_expr(prop, scope))),
                            value: P(read_ssa_to_expr(val, scope)),
                        }),
                        ir::PropKind::Get | ir::PropKind::Set => {
                            unimplemented!("getter and setter props cannot yet be elaborated")
                        }
                    }))
                })
                .collect(),
        }),
        ir::Expr::RegExp {
            regex,
            has_escape,
            flags,
        } => ast::Expr::Lit(ast::Lit::Regex(ast::Regex {
            span: span(),
            exp: ast::Str {
                span: span(),
                value: regex,
                has_escape,
            },
            flags: flags.map(|f| ast::Str {
                span: span(),
                value: f,
                has_escape: false,
            }),
        })),
        ir::Expr::Unary { op, val } => ast::Expr::Unary(ast::UnaryExpr {
            span: span(),
            op: match op {
                ir::UnaryOp::Plus => ast::UnaryOp::Plus,
                ir::UnaryOp::Minus => ast::UnaryOp::Minus,
                ir::UnaryOp::Not => ast::UnaryOp::Bang,
                ir::UnaryOp::Tilde => ast::UnaryOp::Tilde,
                ir::UnaryOp::Typeof => ast::UnaryOp::TypeOf,
                ir::UnaryOp::Void => ast::UnaryOp::Void,
            },
            arg: P(read_ssa_to_expr(val, scope)),
        }),
        ir::Expr::Binary { op, left, right } => ast::Expr::Bin(ast::BinExpr {
            span: span(),
            op: match op {
                ir::BinaryOp::EqEq => ast::BinaryOp::EqEq,
                ir::BinaryOp::NotEq => ast::BinaryOp::NotEq,
                ir::BinaryOp::StrictEq => ast::BinaryOp::EqEqEq,
                ir::BinaryOp::NotStrictEq => ast::BinaryOp::NotEqEq,
                ir::BinaryOp::Lt => ast::BinaryOp::Lt,
                ir::BinaryOp::LtEq => ast::BinaryOp::LtEq,
                ir::BinaryOp::Gt => ast::BinaryOp::Gt,
                ir::BinaryOp::GtEq => ast::BinaryOp::GtEq,
                ir::BinaryOp::ShiftLeft => ast::BinaryOp::LShift,
                ir::BinaryOp::ShiftRight => ast::BinaryOp::RShift,
                ir::BinaryOp::ShiftRightZero => ast::BinaryOp::ZeroFillRShift,
                ir::BinaryOp::Add => ast::BinaryOp::Add,
                ir::BinaryOp::Sub => ast::BinaryOp::Sub,
                ir::BinaryOp::Mul => ast::BinaryOp::Mul,
                ir::BinaryOp::Div => ast::BinaryOp::Div,
                ir::BinaryOp::Mod => ast::BinaryOp::Mod,
                ir::BinaryOp::BitOr => ast::BinaryOp::BitOr,
                ir::BinaryOp::BitXor => ast::BinaryOp::BitXor,
                ir::BinaryOp::BitAnd => ast::BinaryOp::BitAnd,
                ir::BinaryOp::Exp => ast::BinaryOp::Exp,
                ir::BinaryOp::In => ast::BinaryOp::In,
                ir::BinaryOp::Instanceof => ast::BinaryOp::InstanceOf,
            },
            left: P(read_ssa_to_expr(left, scope)),
            right: P(read_ssa_to_expr(right, scope)),
        }),
        ir::Expr::Delete { obj, prop } => ast::Expr::Unary(ast::UnaryExpr {
            span: span(),
            op: ast::UnaryOp::Delete,
            arg: P(ast::Expr::Member(ast::MemberExpr {
                span: span(),
                obj: ast::ExprOrSuper::Expr(P(read_ssa_to_expr(obj, scope))),
                prop: P(read_ssa_to_expr(prop, scope)),
                computed: true,
            })),
        }),
        ir::Expr::Yield { kind, val } => ast::Expr::Yield(ast::YieldExpr {
            span: span(),
            arg: Some(P(read_ssa_to_expr(val, scope))),
            delegate: match kind {
                ir::YieldKind::Single => false,
                ir::YieldKind::Delegate => true,
            },
        }),
        ir::Expr::Await { val } => ast::Expr::Await(ast::AwaitExpr {
            span: span(),
            arg: P(read_ssa_to_expr(val, scope)),
        }),
        ir::Expr::Call { kind, func, args } => {
            let callee = P(read_ssa_to_expr(func, scope));
            let args = args
                .into_iter()
                .map(|(kind, val)| ast::ExprOrSpread {
                    spread: match kind {
                        ir::EleKind::Single => None,
                        ir::EleKind::Spread => Some(span()),
                    },
                    expr: P(read_ssa_to_expr(val, scope)),
                })
                .collect();
            match kind {
                ir::CallKind::Call => ast::Expr::Call(ast::CallExpr {
                    span: span(),
                    callee: ast::ExprOrSuper::Expr(callee),
                    args,
                    type_args: None,
                }),
                ir::CallKind::New => ast::Expr::New(ast::NewExpr {
                    span: span(),
                    callee,
                    args: Some(args),
                    type_args: None,
                }),
            }
        }
        ir::Expr::Function { kind, name, body } => unimplemented!(),
        ir::Expr::CurrentFunction | ir::Expr::Argument { .. } => {
            unreachable!("CurrentFunction and Argument should be removed by convert_stmt")
        }
    }
}

fn read_ssa_to_expr(ssa_ref: ir::Ref<ir::SSA>, scope: &scope::Ir) -> ast::Expr {
    let name = match scope.get_ssa(&ssa_ref) {
        Some(name) => name,
        None => unreachable!("reading from undeclared SSA ref"),
    };
    ast::Expr::Ident(ast::Ident {
        span: span(),
        sym: name,
        type_ann: None,
        optional: false,
    })
}

fn span() -> Span {
    // todo make sourcemaps work by wiring this through from the original AST
    Span::default()
}
