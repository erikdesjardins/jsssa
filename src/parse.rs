use std::env;
use std::fs::File;
use std::process::{Command, Stdio};
use std::io::Write;

use failure::{Error, ResultExt};
use rand::{self, Rng};
use serde_json;

use ast;

static BABYLON_BIN: &[u8] = include_bytes!("../vendor/babylon.js");

pub fn parse(js: &str) -> Result<ast::File, Error> {
    let mut path = env::temp_dir();
    let random_name: String = rand::thread_rng().gen_ascii_chars().take(32).collect();
    path.push(random_name);

    File::create(&path)?
        .write_all(BABYLON_BIN)?;

    let mut child = Command::new("node")
        .arg(&path)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .spawn()?;

    write!(child.stdin.as_mut().unwrap(), "{}", js)?;

    let out = child.wait_with_output()?;

    if !out.status.success() {
        let stderr = String::from_utf8(out.stderr)?;
        // skip error throwing header
        let stderr = stderr
            .lines()
            .skip(3)
            .collect::<Vec<_>>()
            .join("\n");

        return Err(ParseError { stderr })?;
    }

    let stdout = String::from_utf8(out.stdout)?;
    let ast = serde_json::from_str(&stdout).context("failed to parse babylon output")?;

    Ok(ast)
}

#[derive(Fail, Debug)]
#[fail(display = "JS parse error: {}", stderr)]
struct ParseError {
    stderr: String
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn basic_foo() {
        use ast::*;

        assert_eq!(
            parse(
                r#"
                    'use strict';
                    function foo(x) {
                        return x + 1;
                    }
                "#
            ).unwrap(),
            File::new(Program::new(
                vec![
                    FunctionDeclaration::new(
                        Identifier::new("foo".to_owned()),
                        vec![Identifier::new("x".to_owned()).into()],
                        BlockStatement::new(
                            vec![
                                ReturnStatement::new(Some(
                                    BinaryExpression::new(
                                        BinaryOperator::Add,
                                        Box::new(Identifier::new("x".to_owned()).into()),
                                        Box::new(NumericLiteral::new(1.0).into()),
                                    ).into(),
                                )).into(),
                            ],
                            vec![],
                        ),
                        false,
                        false,
                    ).into(),
                ],
                vec![
                    Directive::new(DirectiveLiteral::new("use strict".to_owned())),
                ],
                SourceType::Script,
            ))
        )
    }
}
