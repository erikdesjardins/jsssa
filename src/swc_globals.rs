/// A token proving that SWC's globals are initialized.
pub struct Initialized(());

/// Initializes the globals used by SWC and provides a token as proof.
///
/// A higher-rank lifetime ensures the token cannot escape the closure.
pub fn with<R>(f: impl FnOnce(&'_ Initialized) -> R) -> R {
    swc_common::GLOBALS.set(&swc_common::Globals::new(), || f(&Initialized(())))
}
