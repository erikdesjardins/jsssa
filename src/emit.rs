use std::sync::Arc;

use swc_common::SourceMap;
use swc_ecma_ast as ast;
use swc_ecma_codegen::{text_writer::JsWriter, Config, Emitter, Handlers};

use crate::err::Error;
use crate::swc_globals;

pub struct Opt {
    pub minify: bool,
}

#[inline(never)] // for better profiling
pub fn emit(
    _: &swc_globals::Initialized,
    ast: ast::Script,
    files: Arc<SourceMap>,
    options: Opt,
) -> Result<String, Error> {
    let mut wr = vec![];

    {
        let mut emitter = Emitter {
            cfg: Config {
                minify: options.minify,
            },
            cm: files.clone(),
            wr: Box::new(JsWriter::new(files, "\n", &mut wr, None)),
            comments: None,
            handlers: {
                struct MyHandlers;
                impl Handlers for MyHandlers {}
                Box::new(MyHandlers)
            },
            pos_of_leading_comments: Default::default(),
        };
        emitter.emit_script(&ast)?;
    }

    Ok(String::from_utf8_lossy(&wr).into_owned())
}
