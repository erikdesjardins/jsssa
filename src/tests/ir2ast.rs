use crate::ast2ir;
use crate::emit;
use crate::err::NiceError;
use crate::ir2ast;
use crate::parse;
use crate::swc_globals;

macro_rules! case {
    ( $name:ident, $string:expr ) => {
        #[test]
        fn $name() -> Result<(), NiceError> {
            swc_globals::with(|g| {
                let (ast, files) = parse::parse(g, $string)?;
                let ir = ast2ir::convert(g, ast);
                let ast2 = ir2ast::convert(g, ir);
                let js = emit::emit(g, ast2, files)?;
                insta::assert_snapshot_matches!(stringify!($name), js, $string);
                Ok(())
            })
        }
    };
}

case!(
    basic,
    r#"
    function f(x) {
        while (true);
        x = y.bar;
        z.foo = x ? true : 'hi';
        return +[1 || x, { x }, f + 1, ++g];
    }
    f(1), true;
"#
);

case!(
    deep_scopes,
    r#"
    var x = 1;
    (function() {
        (function() {
            x = 2;
        });
    });
"#
);

case!(
    deconflict,
    r#"
    var x = 1;
    var x$1 = 1;
    (function() {
        var x = 1;
    })
"#
);

case!(
    deconflict_globals,
    r#"
    var x = 1;
    (function() {
        var x = 1;
        x$1;
        x$2;
    })
"#
);

case!(
    deconflict_for_in,
    r#"
    {
        const x;
    }
    for (x in 1);
"#
);

case!(
    empty_blocks,
    r#"
    if (x) {
        good;
    } else {}

    try { good; } catch {} finally {}

    try { good; } catch {} finally { good; }
"#
);

case!(
    labels,
    r#"
    outer: for (;;) {
        inner: for (;;) {
            if (foo) continue inner;
            if (bar) break outer;
        }
    }
"#
);

case!(
    string_has_escape_behavior,
    r#"
    "foo";
    "ba\r";
    "ba\\z";
"#
);

case!(
    regex_has_escape_behavior,
    r#"
    /foo/;
    /bar\./;
    /ba\/z/;
"#
);

case!(
    no_void_0_initializer,
    r#"
    var x;
    let y;
"#
);

case!(
    object_props,
    r#"
    var obj = {
        x: 1,
        ['bad>']: 2,
        0: 7,
    };
    obj.y = 3;
    obj['#bad'] = 4;
    obj.z;
    obj['%bad'];
    delete obj.w;
    delete obj['^bad'];
"#
);
