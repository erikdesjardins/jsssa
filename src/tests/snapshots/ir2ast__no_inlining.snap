---
created: "2019-04-06T23:17:34.700679100Z"
creator: insta@0.7.1
source: src/tests/ir2ast.rs
expression: js
---
var _ini = y;
var x = _ini;

